plugin/distributed/distributed.go

```diff
@@ -94,10 +98,105 @@ const writeResultsAction = "writeResults"
 // Key that results are stored under
 const requestResultKey = "results"
 
-// Just used for unmarshalling the results passed from osquery.
-type resultsStruct struct {
+// OsqueryInt handles unmarshaling integers in noncanonical osquery json.
+type OsqueryInt int
+
+// UnmarshalJSON marshals a json string, that is convertable to an int, for
+// example "234" -> 234.
+func (oi *OsqueryInt) UnmarshalJSON(buff []byte) error {
+	// zero value
+	if string(buff) == `""` {
+		return nil
+	}
+	val, err := strconv.Atoi(string(buff[1 : len(buff)-1]))
+	if err != nil {
+		return &json.UnmarshalTypeError{
+			Value:  string(buff),
+			Type:   reflect.TypeOf(oi),
+			Struct: "statuses",
+		}
+	}
+	*oi = OsqueryInt(val)
+	return nil
+}
+
+// ResultsStruct is used for unmarshalling the results passed from osquery.
+type ResultsStruct struct {
 	Queries  map[string][]map[string]string `json:"queries"`
-	Statuses map[string]string              `json:"statuses"`
+	Statuses map[string]OsqueryInt          `json:"statuses"`
+}
+
+// UnmarshalJSON turns structurally inconsistent osquery json into a ResultsStruct.
+func (rs *ResultsStruct) UnmarshalJSON(buff []byte) error {
+	emptyRow := []map[string]string{}
+	rs.Queries = make(map[string][]map[string]string)
+	rs.Statuses = make(map[string]OsqueryInt)
+	// Queries can be []map[string]string OR an empty string
+	// so we need to deal with an interface to accomodate two types
+	intermediate := struct {
+		Queries  map[string]interface{} `json:"queries"`
+		Statuses map[string]OsqueryInt  `json:"statuses"`
+	}{}
+	if err := json.NewDecoder(bytes.NewBuffer(buff)).Decode(&intermediate); err != nil {
+		return err
+	}
+	for queryName, status := range intermediate.Statuses {
+		rs.Statuses[queryName] = status
+		// sometimes we have a status but don't have a corresponding
+		// result
+		queryResult, ok := intermediate.Queries[queryName]
+		if !ok {
+			rs.Queries[queryName] = emptyRow
+			continue
+		}
+		// deal with structurally inconsistent results, sometimes a query
+		// without any results is just a name with an empty string
+		switch val := queryResult.(type) {
+		case string:
+			rs.Queries[queryName] = emptyRow
+		case []interface{}:
+			results, err := convertRows(val)
+			if err != nil {
+				return err
+			}
+			rs.Queries[queryName] = results
+		default:
+			return fmt.Errorf("results for %q unknown type", queryName)
+		}
+	}
+	return nil
+}
+
+func (rs *ResultsStruct) toResults() ([]Result, error) {
+	var results []Result
+	for queryName, rows := range rs.Queries {
+		var result Result
```

Comment:
Readability nit: Use the `Result{QueryName: queryName}` style of populating a struct rather than `=`.